src/lib/FreeC/Backend/Agda/Converter/TypeDecl.hs

```diff
@@ -61,19 +73,22 @@ convertDataDecl ident@(IR.DeclIdent srcSpan name) typeVars constrs =
 convertConDecl
   :: IR.DeclIdent     -- ^ The identifier of the data type.
   -> [IR.TypeVarDecl] -- ^ The type parameters declared by the data type.
+  -> Bool             -- ^ Is this a recursive data declaration
   -> [IR.ConDecl]     -- ^ The constructor declarations of the data type.
   -> Converter [Agda.Declaration]
-convertConDecl (IR.DeclIdent srcSpan ident) typeVars =
-  mapM $ convertConstructor $ IR.typeApp NoSrcSpan
-                                         (IR.TypeCon srcSpan ident)
-                                         (map IR.typeVarDeclToType typeVars)
+convertConDecl (IR.DeclIdent srcSpan ident) typeVars isRec =
+  mapM $ convertConstructor (bool Nothing (Just ident) isRec) $ IR.typeApp
+    NoSrcSpan
+    (IR.TypeCon srcSpan ident)
+    (map IR.typeVarDeclToType typeVars)
 
 -- | Converts a single constructor of a (non-recursive) data type.
-convertConstructor :: IR.Type -> IR.ConDecl -> Converter Agda.Declaration
-convertConstructor retType (IR.ConDecl _ (IR.DeclIdent srcSpan name) argTypes)
+convertConstructor
+  :: Maybe IR.QName -> IR.Type -> IR.ConDecl -> Converter Agda.Declaration
+convertConstructor ident retType (IR.ConDecl _ (IR.DeclIdent srcSpan name) argTypes)
```

Comment:
This should be named something with `ConDecl` as well (`convertConDecl` should probably be named `convertConDecls` and `this function `convertConDecl`). Add documentation for the new argument. Is this really only used for non-recursive data types?